oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MultithreadedTraverseWithSortStrategy.java

```diff
@@ -408,6 +454,174 @@ public void run() {
+     *      <li>construct new list of files to be merged by checking if its already merged</li>
+     *    and create intermediate merge file
+     *    (if final merge) merge all intermediate merge files and create sorted file
+     *      <li>add all merged files to merged list</li>
+     * </ol>
+     */
+    private class MergeRunner implements Runnable {
+        private final ArrayList<File> mergedFiles = new ArrayList<File>();
+        private final ArrayList<File> unmergedFiles = new ArrayList<File>();
+        private final File sortedFile;
+        private final ExecutorService executorService;
+        private final int threadPoolSize = Integer.getInteger(PROP_MERGE_THREAD_POOL_SIZE, DEFAULT_NUMBER_OF_MERGE_TASK_THREADS);
+        private final int batchMergeSize = Integer.getInteger(PROP_MERGE_TASK_BATCH_SIZE, DEFAULT_NUMBER_OF_FILES_PER_MERGE_TASK);
+        private final Comparator fileSizeComparator = new SizeFileComparator();
+
+        public MergeRunner(File sortedFile) {
+            this.sortedFile = sortedFile;
+            this.executorService = Executors.newFixedThreadPool(threadPoolSize);
+        }
+
+        private List<File> getSmallestUnmergedFiles(int size) {
+            ArrayList<File> result = new ArrayList<File>(unmergedFiles);
+            result.remove(MERGE_POISON_PILL);
+            result.sort(fileSizeComparator);
+            int endIdx = size > result.size() ? result.size() : size;
+            return result.subList(0, endIdx);
+        }
+
+        private void markAsMerged(List<File> target) {
+            mergedFiles.addAll(target);
+            unmergedFiles.removeAll(target);
+        }
+
+        @Override
+        public void run() {
+            Phaser mergeTaskPhaser = new Phaser(1);
+            List<Future<File>> results = new ArrayList<>();
+            List<File> mergeTarget = new ArrayList<>();
+            int count = 0;
+
+            while (true) {
+                try {
+                    unmergedFiles.add(sortedFiles.take());
+                    if (sortedFiles.contains(MERGE_POISON_PILL) || unmergedFiles.contains(MERGE_POISON_PILL)) {
+                        break;
+                    }
+                    // add another batchMergeSize so that we choose the smallest of a larger range
+                    if (unmergedFiles.size() >= 2*batchMergeSize) {
+                        count++;
+                        mergeTarget.clear();
+                        mergeTarget = getSmallestUnmergedFiles(batchMergeSize);
+                        Callable<File> mergeTask = new MergeTask(mergeTarget, mergeTaskPhaser,
+                                new File(mergeDir, String.format("intermediate-%s", count)));
+                        markAsMerged(mergeTarget);
+                        results.add(executorService.submit(mergeTask));
+                    }
+                } catch (InterruptedException e) {
+                    log.error("Failed while draining from sortedFiles {}", e);
```

Comment:
Actually, here the right code would be:

    log.error("Failed while draining from sortedFiles", e);

... this will log the stack trace.

Reply:
make sense